telethon/helpers.py

```diff
@@ -64,40 +64,74 @@ def within_surrogate(text, index, *, length=None):
 
 def strip_text(text, entities):
     """
-    Strips whitespace from the given text modifying the provided entities.
+    Strips whitespace from the given surrogated text modifying the provided
+    entities, also removing any empty (0-length) entities.
 
-    This assumes that there are no overlapping entities, that their length
-    is greater or equal to one, and that their length is not out of bounds.
+    This assumes that the length of entities is greater or equal to 0, and
+    that no entity is out of bounds.
     """
     if not entities:
         return text.strip()
 
-    while text and text[-1].isspace():
-        e = entities[-1]
-        if e.offset + e.length == len(text):
-            if e.length == 1:
-                del entities[-1]
-                if not entities:
-                    return text.strip()
-            else:
-                e.length -= 1
-        text = text[:-1]
-
-    while text and text[0].isspace():
-        for i in reversed(range(len(entities))):
-            e = entities[i]
-            if e.offset != 0:
-                e.offset -= 1
-                continue
-
-            if e.length == 1:
-                del entities[0]
-                if not entities:
-                    return text.lstrip()
+    len_ori = len(text)
+    text = text.lstrip()
+    left_offset = len_ori - len(text)
+    text = text.rstrip()
+    len_final = len(text)
+
+    for i in reversed(range(len(entities))):
+        e = entities[i]
+        if e.length == 0:
+            del entities[i]
+            continue
+
+        if e.offset + e.length > left_offset:
+            if e.offset >= left_offset:
+                #  0 1|2 3 4 5       |       0 1|2 3 4 5
+                #     ^     ^        |          ^
+                #   lo(2)  o(5)      |      o(2)/lo(2)
+                e.offset -= left_offset
+                #     |0 1 2 3       |          |0 1 2 3
+                #           ^        |          ^
+                #     o=o-lo(3=5-2)  |    o=o-lo(0=2-2)
             else:
-                e.length -= 1
+                # e.offset < left_offset and e.offset + e.length > left_offset
+                #  0 1 2 3|4 5 6 7 8 9 10
+                #   ^     ^           ^
+                #  o(1) lo(4)      o+l(1+9)
+                e.length = e.offset + e.length - left_offset
```

Comment:
I personally find it clearer as

```python
e.length -= (left_offset - e.offset)
```

But this works too.

Reply:
Also note that this might set the `length` to `0`. Imagine:

* `offset = 0`
* `left_offset = 4`
* `length = 4`

So it would be preferable to move the `length 0` check to the end, I think.